### PR TITLE
Add support for async client headers

### PR DESCRIPTION
**Pull request by alaister**
Hello,
This is a feature I need when working on a react native app using expo secure store, which has an async set/get interface.

At this stage this PR is more a proof of concept with the most simple of changes to get it working.

I am happy to change the implementation to suit your preferred style (for example maybe changing `getHeadersFn()` to always return a promise instead of handling it at the `httpRequest` level).
Also I am happy to add some tests/docs if you like this idea.

Cheers!

### CODE REVIEW

**Review by KATT — CHANGES_REQUESTED** [review]
Add tests otherwise 👍

File: packages/client/src/internals/TRPCClient.ts [hunk 1]
```diff
@@ -48,7 +48,7 @@ export type CreateTRPCClientOptions<TRouter extends AnyRouter> = {
   /**
    * headers to be set on outgoing requests / callback that of said headers
    */
-  headers?: HttpHeaders | (() => HttpHeaders);
+  headers?: HttpHeaders | (() => HttpHeaders) | (() => Promise<HttpHeaders>);
```
Comment (KATT) [inline packages/client/src/internals/TRPCClient.ts:51]: Tiny tweak:

```suggestion
  headers?: HttpHeaders | (() => HttpHeaders | Promise<HttpHeaders>)
```

**Comment by alaister** [comment]
Are there any existing files for testing `httpRequest`? Or should I create a new one?
Also any specific types of tests you'd like?

File: packages/client/src/links/core.ts [hunk 2]
```diff
@@ -43,7 +43,7 @@ export interface HttpLinkOptions {
 export type HttpHeaders = Record<string, string | string[] | undefined>;
 export type LinkRuntimeOptions = Readonly<{
   transformer: DataTransformer;
-  headers: () => HttpHeaders;
+  headers: (() => HttpHeaders) | (() => Promise<HttpHeaders>);
```
Comment (KATT) [inline packages/client/src/links/core.ts:46]:
```suggestion
  headers: (() => HttpHeaders | Promise<HttpHeaders>);
```

**Review by KATT — COMMENTED** [review]
tiny tweaks here

also added you as a contributor, please make a branch based on origin so we can run tests, etc

https://github.com/trpc/trpc/invitations

**Comment by KATT** [comment]
> Are there any existing files for testing `httpRequest`? Or should I create a new one?
> Also any specific types of tests you'd like?

See https://github.com/trpc/trpc/pull/903

**Review by KATT — CHANGES_REQUESTED** [review]
This looks good but please push your branch to the trpc origin and open up a new PR from that instead of your fork

**Comment by alaister** [comment]
Moved to #909

**Pull request merged**